### PR TITLE
Incorporate availability zone into placement decision

### PR DESCRIPTION
**Pull request by spangenberg**

### CODE REVIEW

**Comment by ingvagabund** [comment]
```
I1102 08:41:07.239436       1 actuator.go:110] creating machine
I1102 08:41:07.286589       1 actuator.go:187] no stopped instances found for machine pr-309-97-machine-06e282
I1102 08:41:07.286618       1 actuator.go:301] Describing AMI based on filters
I1102 08:41:09.421021       1 actuator.go:216] Describing security groups based on filters
I1102 08:41:09.587977       1 actuator.go:247] Describing subnets based on filters
W1102 08:41:09.939633       1 actuator.go:345] More than one subnet id returned, only first one will be used
E1102 08:41:10.165161       1 actuator.go:429] error creating EC2 instance: InvalidParameterValue: Value (us-east-1a) for parameter availabilityZone is invalid. Subnet 'subnet-0319201115502e4ee' is in the availability zone us-east-1c
	status code: 400, request id: 3c5c94c7-371e-4b99-a594-8fc59a67a007
E1102 08:41:10.165223       1 actuator.go:113] error creating machine: error creating EC2 instance: InvalidParameterValue: Value (us-east-1a) for parameter availabilityZone is invalid. Subnet 'subnet-0319201115502e4ee' is in the availability zone us-east-1c
	status code: 400, request id: 3c5c94c7-371e-4b99-a594-8fc59a67a007
```

**Comment by ingvagabund** [comment]
All the testing machine sets [1]:
```
Placement: providerconfigv1.Placement{
	Region:           "us-east-1",
	AvailabilityZone: "us-east-1a",
},
```

We will need to modify the TF so it creates all the resources only in a single AZ [2]. Or, later the machine manifests to set either the subnet or the availability zone but not both.

[1] https://github.com/openshift/cluster-api-provider-aws/blob/master/test/utils/manifests.go#L53
[2] https://github.com/openshift/cluster-api-provider-aws/tree/master/hack/prebuild

File: pkg/cloud/aws/actuators/machine/actuator_test.go [hunk 1]
```diff
@@ -412,3 +412,132 @@ func TestBuildEC2Filters(t *testing.T) {
...
+				},
+			},
+		}, nil)
+}
+
+type placementMatcher struct {
+	x *ec2.Placement
+}
+
+func (m placementMatcher) Matches(x interface{}) bool {
```
Comment (enxebre) [inline pkg/cloud/aws/actuators/machine/actuator_test.go]: is this ever called?

File: pkg/cloud/aws/actuators/machine/actuator_test.go [hunk 2]
```diff
@@ -412,3 +412,132 @@ func TestBuildEC2Filters(t *testing.T) {
...
+			subnet:           "subnet-b46032ec",
+		},
+	}
+
+	codec, err := providerconfigv1.NewCodec()
+	if err != nil {
+		t.Fatalf("unable to build codec: %v", err)
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
```
Comment (enxebre) [inline pkg/cloud/aws/actuators/machine/actuator_test.go]: what is this actually validating, I couldn't find any assertion?

**Comment by enxebre** [comment]
- @ingvagabund @spangenberg Terraform should create subnets with filters across all az (it should be already doing this) - this is a real valid case.
- Then the actuator should be able to know what to do for any possible case combining subnetID, subnet filters, and az (This is almost all done in this PR).
- We just need  now to retrieve the subnet by AZ here https://github.com/openshift/cluster-api-provider-aws/blob/master/pkg/cloud/aws/actuators/machine/actuator.go#L340-L353 instead of returning always the first one
- Finally we want to validate this behaviour so the instance is running in the expected AZ here https://github.com/openshift/cluster-api-provider-aws/blob/master/test/machines/machines_test.go#L142

File: pkg/cloud/aws/actuators/machine/actuator_test.go (hunk 1 above)
Comment (spangenberg) [reply pkg/cloud/aws/actuators/machine/actuator_test.go]: Yes from line 508:

`	mockAWSClient.EXPECT().RunInstances(Placement(placement)).Return(`

`placementMatcher` implements `gomock.Matcher`

File: pkg/cloud/aws/actuators/machine/actuator_test.go (hunk 2 above)
Comment (spangenberg) [reply pkg/cloud/aws/actuators/machine/actuator_test.go]: It asserts the placement trough the mock assertion, line 487.

File: pkg/cloud/aws/actuators/machine/actuator_test.go [hunk 3]
```diff
@@ -412,3 +412,132 @@ func TestBuildEC2Filters(t *testing.T) {
...
+						Code: aws.Int64(16),
+					},
+					LaunchTime: aws.Time(time.Now()),
+				},
+			},
+		}, nil)
+}
+
+type placementMatcher struct {
+	x *ec2.Placement
```
Comment (enxebre) [inline pkg/cloud/aws/actuators/machine/actuator_test.go]: x?

File: pkg/cloud/aws/actuators/machine/actuator.go [hunk 4]
```diff
@@ -237,16 +237,20 @@ func getSecurityGroupsIDs(securityGroups []providerconfigv1.AWSResourceReference
...
+			filters = append(filters, providerconfigv1.Filter{Name: "availabilityZone", Values: []string{availabilityZone}})
+		}
+		filters = append(filters, subnet.Filters...)
 		glog.Info("Describing subnets based on filters")
 		describeSubnetRequest := ec2.DescribeSubnetsInput{
-			Filters: buildEC2Filters(subnet.Filters),
+			Filters: buildEC2Filters(filters),
 		}
 		describeSubnetResult, err := client.DescribeSubnets(&describeSubnetRequest)
 		if err != nil {
```
Comment (enxebre) [inline pkg/cloud/aws/actuators/machine/actuator.go]: I know it was not introduced in this PR but would you mind to update copy/paster security groups message in lines 257/258?

Comment (spangenberg) [reply pkg/cloud/aws/actuators/machine/actuator.go]: Sure, done :)

File: pkg/cloud/aws/actuators/machine/actuator_test.go (hunk 3 above)
Comment (spangenberg) [reply pkg/cloud/aws/actuators/machine/actuator_test.go]: Improved it after retouching code anyhow.

**Pull request merged**